src/coreclr/vm/comsynchronizable.cpp

```diff
@@ -703,23 +621,23 @@ void ReleaseThreadExternalCount(Thread * pThread)
 typedef Holder<Thread *, DoNothing, ReleaseThreadExternalCount> ThreadExternalCountHolder;
 
 // Wait for the thread to die
-BOOL ThreadNative::DoJoin(THREADBASEREF DyingThread, INT32 timeout)
+static BOOL DoJoin(THREADBASEREF dyingThread, INT32 timeout)
 {
     CONTRACTL
     {
         THROWS;
         GC_TRIGGERS;
         MODE_COOPERATIVE;
-        PRECONDITION(DyingThread != NULL);
+        PRECONDITION(dyingThread != NULL);
         PRECONDITION((timeout >= 0) || (timeout == INFINITE_TIMEOUT));
     }
     CONTRACTL_END;
 
-    Thread * DyingInternal = DyingThread->GetInternal();
+    Thread* DyingInternal = dyingThread->GetInternal();
```

Comment:
I have looked into the lifetime guarantees of managed and native `Thread`. It turns out that switching to cooperative mode is not sufficient to make this robust against resurrected threads. Opened https://github.com/dotnet/runtime/issues/107473

Switching to cooperative mode make the use-after-free crash harder to hit, but it does not eliminate the window,

Reply:
BTW: https://github.com/dotnet/runtime/pull/107468 is another rare race condition introduced by managed / native Thread interplay that just caught up with us. These issues always come in related batches.